File: java/src/Bj14889.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Bj14889 {
    private static int[][] arr;
    private static boolean[] visit;
    private static int N = 0;
    private static int ans = Integer.MAX_VALUE;

    public static void run() {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        try {
            N = Integer.parseInt(br.readLine());
            arr = new int[N][N];
            visit = new boolean[N];
            for (int i = 0; i < N; i++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                for (int j = 0; j < N; j++) {
                    arr[i][j] = Integer.parseInt(st.nextToken());
                }
            }
            solution(0, 0);
            System.out.println(ans);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void solution(int start, int depth) {
        if (depth == N / 2) {
            diff();
            return;
        }
        for (int i = start; i < N; i++) {
            if (visit[i]) continue;
            visit[i] = true;
            solution(i + 1, depth + 1);
            visit[i] = false;
        }
    }

    private static void diff() {
        int scoreA = 0;
        int scoreB = 0;
        for (int i = 0; i < N - 1; i++) {
            for (int j = i + 1; j < N; j++) {
                if (visit[i] && visit[j]) {
                    scoreA += arr[i][j] + arr[j][i];
                } else if (!visit[i] && !visit[j]) {
                    scoreB += arr[i][j] + arr[j][i];
                }
            }
        }
        int dif = Math.abs(scoreA - scoreB);
        ans = Math.min(ans, dif);
        if (ans == 0) {
            System.out.println(ans);
            System.exit(0);
        }
    }
}

File: kotlin/src/Bj14889.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

class Bj14889 {
    companion object {
        private var ans = Integer.MAX_VALUE;
        private var br = BufferedReader(InputStreamReader(System.`in`))
        private val N = br.readLine().toInt()
        private var arr = Array(N) { Array(N) { 0 } }
        private var visit = Array<Boolean>(N, { false })

        fun solve() {
            for (i in 0..N - 1) {
                var st = StringTokenizer(br.readLine());
                for (j in 0..N - 1) {
                    arr[i][j] = st.nextToken().toInt()
                }
            }
            solution(0, 0)
            println(ans)
        }

        private fun solution(start: Int, depth: Int) {
            if (depth == N / 2) {
                diff()
                return
            }
            for (i in start..N - 1) {
                if (visit[i]) continue
                visit[i] = true
                solution(i + 1, depth + 1)
                visit[i] = false
            }
        }


        private fun diff() {
            var scoreA = 0
            var scoreB = 0
            for (i in 0..N - 2) {
                for (j in i + 1..N - 1) {
                    if (visit[i] && visit[j]) {
                        scoreA += arr[i][j] + arr[j][i]
                    } else if (!visit[i] && !visit[j]) {
                        scoreB += arr[i][j] + arr[j][i]
                    }
                }
            }
            val dif = Math.abs(scoreA - scoreB)
            ans = Math.min(ans, dif)
            if (ans == 0) {
                println(ans)
                System.exit(0);
            }
        }
    }
}